src/main/java/org/folio/circulation/domain/CreateRequestService.java

```diff
@@ -19,65 +18,66 @@
 import org.folio.circulation.support.http.server.ValidationError;
 
 public class CreateRequestService {
+
   private final RequestRepository requestRepository;
   private final UpdateItem updateItem;
   private final UpdateLoanActionHistory updateLoanActionHistory;
   private final RequestPolicyRepository requestPolicyRepository;
+  private final LoanRepository loanRepository;
   private final UpdateLoan updateLoan;
 
-  public CreateRequestService(
-    RequestRepository requestRepository,
-    UpdateItem updateItem,
-    UpdateLoanActionHistory updateLoanActionHistory,
-    UpdateLoan updateLoan,
-    RequestPolicyRepository requestPolicyRepository) {
+  public CreateRequestService(RequestRepository requestRepository, UpdateItem updateItem,
+      UpdateLoanActionHistory updateLoanActionHistory, UpdateLoan updateLoan,
+      RequestPolicyRepository requestPolicyRepository, LoanRepository loanRepository) {
 
     this.requestRepository = requestRepository;
     this.updateItem = updateItem;
     this.updateLoanActionHistory = updateLoanActionHistory;
     this.updateLoan = updateLoan;
     this.requestPolicyRepository = requestPolicyRepository;
+    this.loanRepository = loanRepository;
   }
 
   public CompletableFuture<Result<RequestAndRelatedRecords>> createRequest(
-    RequestAndRelatedRecords requestAndRelatedRecords) {
-
-    return completedFuture(refuseWhenItemDoesNotExist(requestAndRelatedRecords)
-      .next(CreateRequestService::refuseWhenInvalidUserAndPatronGroup)
-      .next(CreateRequestService::refuseWhenItemIsNotValid)
-      .next(CreateRequestService::refuseWhenUserHasAlreadyRequestedItem))
-      .thenComposeAsync(r -> r.after(requestPolicyRepository::lookupRequestPolicy))
-      .thenApply(r -> r.next(CreateRequestService::refuseWhenRequestCannotBeFulfilled))
-      .thenApply(r -> r.map(CreateRequestService::setRequestQueuePosition))
-      .thenComposeAsync(r -> r.after(updateItem::onRequestCreation))
-      .thenComposeAsync(r -> r.after(updateLoanActionHistory::onRequestCreation))
-      .thenComposeAsync(r -> r.after(updateLoan::onRequestCreation))
-      .thenComposeAsync(r -> r.after(requestRepository::create));
+      RequestAndRelatedRecords requestAndRelatedRecords) {
+
+    return findOpenLoanWithRequest(requestAndRelatedRecords, loanRepository)
```

Comment:
This does make it easier to find the open loan. 

I think it means the overall flow is more complicated (the nesting being an indication)

This approach has been used in other places in the code. I think it might be a symptom of a limitation of making decisions beyond only composing asynchronous chunks together.

Would it be useful if I pushed a branch with one of the alternatives I proposed? Or were neither particularly useful?

Reply:
I don't want to create extra work for you but I am sure such a branch would be instructive.

Reply:
I had a really quick play at an alternative (it is still a long way from what I think it could be like, but that is way beyond this issue)

https://github.com/folio-org/mod-circulation/tree/CIRC-258-experiment